modules/adotBidAdapter.js

```diff
@@ -0,0 +1,636 @@
+    ad: creative.markup,
+    adUrl: creative.markupUrl,
+    vastXml: null,
+    vastUrl: null,
+    renderer: creative.renderer
+  };
+}
+
+function getVastXml(ad, done) {
+  if (isString(ad.adUrl)) {
+    window.fetch(ad.adUrl)
+      .then(response => response.text())
+      .then(vastXml => done(null, vastXml))
+      .catch(err => done(err, null));
+
+    return;
+  }
+
+  return done(null, ad.ad);
+}
+
+function formatNativeData(adm) {
+  const parsedAdm = tryParse(adm);
+  const {assets, link: {url, clicktrackers}, imptrackers, jstracker} = parsedAdm.native;
+  const placements = NATIVE_PLACEMENTS;
+  const placementIds = NATIVE_ID_MAPPING;
+
+  return assets.reduce((acc, asset) => {
+    const placementName = placementIds[asset.id];
+    const content = placementName && asset[placements[placementName].name];
+    if (!content) return acc;
+    acc[placementName] = content.text || content.value || {url: content.url, width: content.w, height: content.h};
+    return acc;
+  }, {
+    clickUrl: url,
+    clickTrackers: clicktrackers,
+    impressionTrackers: imptrackers,
+    javascriptTrackers: jstracker && [jstracker]
+  });
+}
+
+function getSizeFromBid(bid, impressionData) {
+  if (isNumber(bid.w) && isNumber(bid.h)) {
+    return { width: bid.w, height: bid.h };
+  }
+
+  if (isImpressionAVideo(impressionData.openRTB)) {
+    const { video } = impressionData.openRTB;
+
+    if (isNumber(video.w) && isNumber(video.h)) {
+      return { width: video.w, height: video.h };
+    }
+  }
+
+  return { width: null, height: null };
+}
+
+function getCreativeFromBid(bid, impressionData) {
+  const shouldUseAdMarkup = !!bid.adm;
+
+  return {
+    markup: shouldUseAdMarkup ? bid.adm : null,
+    markupUrl: !shouldUseAdMarkup ? bid.nurl : null,
+    renderer: getRendererFromBid(bid, impressionData)
+  };
+}
+
+function getRendererFromBid(bid, impressionData) {
+  const isOutstreamImpression = isBidAVideo(bid) &&
+    isImpressionAVideo(impressionData.openRTB) &&
+    isOutstream(impressionData.openRTB.video);
+
+  return isOutstreamImpression
+    ? buildOutstreamRenderer(impressionData)
+    : null;
+}
+
+function buildOutstreamRenderer(impressionData) {
+  const renderer = Renderer.install({
+    url: OUTSTREAM_VIDEO_PLAYER_URL,
+    loaded: false,
+    adUnitCode: impressionData.internal.adUnitCode
+  });
+
+  renderer.setRender((ad) => {
+    getVastXml(ad, (err, vastXml) => {
+      if (err) return;
+
+      ad.renderer.push(() => {
+        const container = impressionData.internal.container
+          ? document.querySelector(impressionData.internal.container)
+          : document.getElementById(impressionData.internal.adUnitCode);
+
+        const player = new window.VASTPlayer(container);
+
+        return player
+          .loadXml(vastXml)
+          .then(() => {
+            player.adVolume = 0;
+          })
+          .then(() => player.startAd())
+          .catch(err => console.error(err));
+      });
+    });
```

Comment:
Similar to the earlier comment, I'm afraid we don't support Promises in Prebid.js.  Can you please rewrite this code in a different way?